datadog-profiling/src/serializer/compressed_streaming_encoder.rs

```diff
@@ -71,28 +78,40 @@ impl CompressedProtobufSerializer {
     pub fn finish(self) -> io::Result<Vec<u8>> {
         match self.compressor {
             Compressor::None => Ok(self.buffer),
-            Compressor::Lz4 { zipper } => {
+            Compressor::Lz4 { encoder: zipper } => {
                 debug_assert!(self.buffer.is_empty());
                 Ok(zipper.finish()?)
             }
+            Compressor::Zstd { encoder } => encoder.finish(),
         }
     }
 
-    pub fn with_config_and_capacity(config: UploadCompression, capacity: usize) -> Self {
+    pub fn with_config_and_capacity(
+        config: UploadCompression,
+        capacity: usize,
+    ) -> io::Result<Self> {
+        const TEMPORARY_BUFFER_SIZE: usize = 256;
         // Final output buffer.
         let buffer = Vec::with_capacity(capacity);
-        match config {
+        Ok(match config {
             UploadCompression::Off => Self {
                 buffer,
                 compressor: Compressor::None,
             },
             UploadCompression::On | UploadCompression::Lz4 => Self {
                 // Temporary input buffer.
-                buffer: Vec::with_capacity(256),
+                buffer: Vec::with_capacity(TEMPORARY_BUFFER_SIZE),
                 compressor: Compressor::Lz4 {
-                    zipper: FrameEncoder::new(buffer),
+                    encoder: Lz4FrameEncoder::new(buffer),
                 },
             },
-        }
+            UploadCompression::Zstd => Self {
+                buffer: Vec::with_capacity(TEMPORARY_BUFFER_SIZE),
+                compressor: Compressor::Zstd {
+                    // A level of 0 uses zstd's default (currently 3).
```

Comment:
Is that the level 3 from the upstream zstd C library? If yes, level 3 is good. I'm asking because there are different zstd implementations out there which assign different meanings to different levels.

Reply:
Yes, at least as far as I understood it.